src/sensai/vector_model.py

```diff
@@ -133,20 +127,113 @@ def withName(self, name: str):
         self.setName(name)
         return self
 
+    def getInputTransformer(self, cls: Type[DataFrameTransformer]):
+        for it in self._inputTransformerChain.dataFrameTransformers:
+            if isinstance(it, cls):
+                return it
+        return None
+
+    def setName(self, name):
+        self._name = name
+
+    def getName(self):
+        if self._name is None:
+            return "unnamed-%s-%x" % (self.__class__.__name__, id(self))
+        return self._name
+
+    def setFeatureGenerator(self, featureGenerator: Optional[FeatureGenerator]):
+        self.withFeatureGenerator(featureGenerator)
+
+    def getFeatureGenerator(self) -> Optional[FeatureGenerator]:
+        return self._featureGenerator
+
+
+class FittableModel(PredictorModel, ABC):
     @abstractmethod
-    def isRegressionModel(self) -> bool:
+    def fit(self, X: pd.DataFrame, Y: pd.DataFrame = None):
+        pass
+
+
+class RulesBasedModel(FittableModel, ABC):
+    """
+    Base class for models where the essential logic is rule based but the feature generators
+    and input transformers can be fit on data frames
+    """
+    def __init__(self):
+        super().__init__()
+        self._isFitted = False
+
+    def fit(self, X: pd.DataFrame, Y: pd.DataFrame = None):
+        self._featureGenerator.fit(X, Y=Y, ctx=self)
+        self._inputTransformerChain.fit(X)
+        self._isFitted = True
+
+    def predict(self, x: pd.DataFrame) -> pd.DataFrame:
+        """
+        Performs a prediction for the given input data frame
+
+        :param x: the input data
+        :return: a DataFrame with the same index as the input
+        """
+        if not self._isFitted:
+            log.warning(
```

Comment:
We could avoid the warning in cases where there is no data transformation and no feature generation.
I am on the fence as to whether in cases where there are, this should be an exception; the logged warning could remain unnoticed.

Reply:
I don't agree here, it will be common that a RuleBasedModel only has RuleBasedFgen and DFTs. Fitting does nothing then and the user shoudn't be forced to do that. In any case for rule based models there is (necessarily) less safety and the user has to be a bit more aware of what they are doing

Reply:
As a user, I would object to receiving warnings that do not apply. Ideally, we should not warn that there might be a problem but _know_ if there is a problem (by having the components be aware of the fitting requirement) - and throw an exception iff so.

Reply:
Maybe this is material for an improvement that adds the "knowledge" to DFTs, etc.

Reply:
I agree. Let's maybe address it in a separate issue? I don't want to further blow up this PR

Reply:
Yes, separate issue.